tdrs-backend/tdpservice/email/email.py

```diff
@@ -0,0 +1,92 @@
+and a dictionary of context variables to be used in the email template. The context vlaues can be found in the
+templates directory.
+
+There are optimizations that can be made to this module for sending mass emails.
+Django's send_mass_mail function can be used to send multiple emails at once.
+https://docs.djangoproject.com/en/4.1/topics/email/#send-mass-mail
+"""
+
+from celery import shared_task
+from django.core.exceptions import ValidationError
+from django.core.validators import validate_email
+from django.core.mail import EmailMultiAlternatives
+from django.conf import settings
+from django.template.loader import get_template
+from django.template import TemplateDoesNotExist
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+@shared_task
+def mail(email_path: str, recipient_email: str, email_context: dict = None) -> None:
+    """Send email to user."""
+    subject = email_context['subject']
+    html_message = construct_email(email_path, email_context)
+
+    if email_context['text_message']:
+        text_message = email_context['text_message']
+    else:
+        text_message = 'An email was sent with HTML content. Please view in an HTML capable email client.'
+    send_email(subject, text_message, html_message, [recipient_email])
+
+def construct_email(email_path: str, context: dict):
+    """Get email template."""
+    try:
+        template = get_template(email_path)
+        return template.render(context)
+    except TemplateDoesNotExist as exc:
+        raise TemplateDoesNotExist(f"Template {email_path} does not exist") from exc
+
+def send_email(subject: str, message: str, html_message: str, recipient_list: list) -> bool:
+    """Send an email to a list of recipients."""
+    valid_emails = validate_emails(recipient_list)
+
+    msg = EmailMultiAlternatives(subject, message, settings.DEFAULT_FROM_EMAIL, valid_emails)
+    msg.attach_alternative(html_message, "text/html")
+    response = msg.send()
+
+    if response == 0:
+        raise Exception(f"Emails were attempted to the following email list: {valid_emails}. \
+        But none were sent. They may be invalid.")
+
+    logger.info('Email sent successfully')
+    return True
+
+def validate_emails(emails: list) -> list:
+    """Validate email addresses."""
+    valid_emails = []
+    for email in emails:
+        if validate_single_email(email):
+            valid_emails.append(email)
+    if len(valid_emails) == 0:
+        raise ValidationError("No valid emails provided.")
+    return valid_emails
+
+def validate_single_email(email: str) -> bool:
+    """Validate email address."""
+    try:
+        validate_email(email)
+        return True
+    except ValidationError:
+        logger.info(f"{email} is not a valid email address. An email will not be sent to this address.")
+        return False
+
+def validate_sender_email(email: str) -> bool:
+    """Validate sender email address."""
```

Comment:
Why is this 3 functions? For testing? 

This feels like it could be a relatively simple function. From what I see, we only call `validate_emails()` in the other mail functions. We raise validation errors multiple times -- I think this can and should be condensed.

Reply:
I consolidated two of those validation functions and am now using the validate_sender_email. I'm not sure validate_sender_email is needed since the sender email will probably always use the value that is hardcoded in the settings.